src/main.cpp

```diff
@@ -90,87 +91,197 @@ namespace
+					if (config->count("admin_only") && config->at("admin_only").get<bool>()) {
+						return user_is_administrator(*rei.rsComm);
+					}
+
+					namespace adm = irods::experimental::administration;
+
+					const adm::user user{rei.uoic->userName, rei.uoic->rodsZone};
+
+					for (auto&& editor : config->at("editors")) {
+						if (const auto& type = editor.at("type").get_ref<const std::string&>(); type == "group") {
+							const adm::group group{editor.at("name").get_ref<const std::string&>()};
+
+							if (adm::server::user_is_member_of_group(*rei.rsComm, group, user)) {
+								return CODE(RULE_ENGINE_CONTINUE);
+							}
+						}
+						else if (type == "user") {
+							if (editor.at("name").get_ref<const std::string&>() ==
+							    adm::server::local_unique_name(*rei.rsComm, user)) {
+								return CODE(RULE_ENGINE_CONTINUE);
+							}
+						}
+					}
+
+					// At this point, the user is not an administrator and they aren't a member of
+					// the editors list. Therefore, we return an error because the user is attempting to
+					// modify metadata in a guarded namespace.
+					return ERROR(CAT_INSUFFICIENT_PRIVILEGE_LEVEL, "User is not allowed to modify metadata");
+				}
+			}
+		}
+		catch (const json::exception&) {
+			// clang-format off
             log_re::error({{"log_message", "Unexpected JSON access or type error."},
                            {"rule_engine_plugin", "metadata_guard"}});
             // clang-format on
-        }
-        catch (const std::exception& e) {
-            log_re::error({{"log_message", e.what()}, {"rule_engine_plugin", "metadata_guard"}});
-        }
-
-        return CODE(RULE_ENGINE_CONTINUE);
-    }
+		}
+		catch (const std::exception& e) {
+			log_re::error({{"log_message", e.what()}, {"rule_engine_plugin", "metadata_guard"}});
+		}
+		return CODE(RULE_ENGINE_CONTINUE);
+	}
+
+	auto pep_atomic_metadata_handler(std::list<boost::any>& _rule_arguments, irods::callback _effect_handler)
+		-> irods::error
+	{
+		auto is_input_valid = [](const bytesBuf_t* _input) -> std::tuple<bool, std::string> {
+			if (!_input) {
+				return {false, "Missing JSON input"};
+			}
+
+			if (_input->len <= 0) {
+				return {false, "Length of buffer must be greater than zero"};
+			}
+
+			if (!_input->buf) {
+				return {false, "Missing input buffer"};
+			}
+
+			return {true, ""};
+		};
+
+		try {
+			auto* input_bb = boost::any_cast<bytesBuf_t*>(*std::next(std::begin(_rule_arguments), 2));
+			auto& rei = get_rei(_effect_handler);
+			const auto config = load_plugin_config(rei);
+
+			if (const auto [valid, msg] = is_input_valid(input_bb); !valid) {
+				log_re::error(msg);
+				return ERROR(INPUT_ARG_NOT_WELL_FORMED_ERR, msg);
+			}
+
+			const auto input_bb_casted = static_cast<const char*>(input_bb->buf);
+			json input = json::parse(input_bb_casted, input_bb_casted + input_bb->len);
+
+			const auto& operations = input.at("operations");
```

Comment:
Kind of a minor point because I know it'll be handled in the `std::exception` catch below, but I think this will throw a different `nlohmann::json::exception` other than `parse_error`. Might need to handle `out_of_range` as well, I'd guess.

https://json.nlohmann.me/api/basic_json/at/#exceptions

Reply:
If the explicit exception types don't lead to changes in the behavior of the plugin, perhaps we collapse all JSON exception types to `const json::exception&`.

Reply:
That sounds reasonable to me, yep.